src/simsopt/solve/mpi.py

```diff
@@ -232,3 +235,276 @@ def _f_proc0(x):
+
+        del_t = time() - start_time
+        objective_file.write(f"{n_objective_evals:6d},{del_t:12.4e}")
+        for xj in x:
+            objective_file.write(f",{xj:24.16e}")
+        objective_file.write(f",{objective_val:24.16e}\n")
+        objective_file.flush()
+
+        n_objective_evals += 1
+        logger.debug(f"objective is {objective_val}")
+        return objective_val
+
+    # wrap the constraints for logging
+    def _nlc_proc0(x):
+        """
+        This function is used for constrained_mpi_solve.  It is called only by
+        proc 0 while workers are in the worker loop.
+        """
+        logger.debug("Entering _nlc_proc0")
+        mpi.mobilize_workers(CALCULATE_NLC)
+        # Send workers the state vector:
+        mpi.comm_groups.bcast(x, root=0)
+        logger.debug("Past bcast in _nlc_proc0")
+
+        try:
+            constraint_val = prob.nonlinear_constraints(x)
+            logger.debug(f"constraints in _nlc_proc0:\n {constraint_val}")
+        except:
+            constraint_val = np.full(prob.nvals, 1.0e12)
+            logger.info("Exception caught during function evaluation.")
+
+        nonlocal constraint_datalog_started, constraint_file, n_constraint_evals
+
+        # Since the number of terms is not known until the first
+        # evaluation of the objective function, we cannot write the
+        # header of the output file until this first evaluation is
+        # done.
+        if not constraint_datalog_started:
+            # Initialize log file
+            constraint_datalog_started = True
+            datestr = datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
+            constraint_file = open(f"constraint_{datestr}.dat", 'w')
+            constraint_file.write(f"Problem type:\nconstrained\nnparams:\n{prob.dof_size}\n")
+            constraint_file.write("function_evaluation,seconds")
+
+            for j in range(prob.dof_size):
+                constraint_file.write(f",x({j})")
+            constraint_file.write(",constraint_function\n")
+            for j in range(len(constraint_val)):
+                constraint_file.write(f",F({j})")
+            constraint_file.write("\n")
+
+        del_t = time() - start_time
+        constraint_file.write(f"{n_constraint_evals:6d},{del_t:12.4e}")
+        for xj in x:
+            constraint_file.write(f",{xj:24.16e}")
+        for fj in constraint_val:
+            constraint_file.write(f",{fj:24.16e}")
+        constraint_file.write("\n")
+        constraint_file.flush()
+
+        n_constraint_evals += 1
+        logger.debug(f"constraints are {constraint_val}")
+        return constraint_val
+
+    # prepare bounds 
+    if prob.has_bounds:
+        bounds = Bounds(prob.lb, prob.ub)
+    else:
+        bounds = None
+
+    # prepare linear constraints
+    constraints = []
+    if prob.has_lc:
+        lincon = LinearConstraint(prob.A_lc, lb=-np.inf, ub=prob.b_lc)
+        constraints.append(lincon)
+
```

Comment:
Moving this block of code inside of ConstrainedProblem poses a slight challenge. If `lincon` is defined in the `__init__` block of ConstrainedProblem, but a user changed `A_lc` after initializing ConstrianedProblem, then `lincon` wont be up-to-date when the problem is solved. Because of this we need to either make sure that the linear constraints are not changed after the initialization of the ConstrainedProblem, or that scipy version of the constraints `lincon` is updated every time the linear constraints change.